server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -44,48 +49,125 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
+  private def toTruncatedInputStream(streamAndFile: PartStream, stopAt: Int): InputStream = streamAndFile match {
+    case ByteStream(stream) =>
+      val bytes = stream.toByteArray
+      new ByteArrayInputStream(bytes.slice(0, stopAt))
+    case FileStream(f, fileOutputStream) =>
+      fileOutputStream.getChannel.truncate(stopAt.toLong)
+      fileOutputStream.close()
+      new FileInputStream(f)
+  }
+
+  private def readUntilNewline(inputStream: InputStream): Array[Byte] =
     Iterator
-      .continually(reader.readLine())
-      .takeWhile(_ != null)
-      .foldLeft(initialParseState) { case (state, line) =>
-        state.parseState match {
-          case Default if line.startsWith(boundary)           => state.changeState(AfterBoundary)
-          case Default                                        => state
-          case AfterBoundary if line.trim.isEmpty             => state.changeState(AfterHeaderSpace)
-          case AfterBoundary                                  => state.addHeader(line)
-          case AfterHeaderSpace if !line.startsWith(boundary) => state.addBody(line.getBytes())
-          case AfterHeaderSpace                               => state.completePart()
-        }
+      .continually(inputStream.read())
+      .sliding(2)
+      .takeWhile(twoChars => twoChars != Seq('\r', '\n') && twoChars != Seq('\n') && !twoChars.lastOption.contains(-1))
+      .map(_.head.toByte)
+      .toArray
+
+  private def readStringUntilNewline(inputStream: InputStream) = {
+    val bytes = readUntilNewline(inputStream)
+    new String(bytes)
+  }
+
+  def parseMultipartBody(inputStream: InputStream, boundary: String): Seq[ParsedMultiPart] = {
+    val boundaryBytes = boundary.getBytes
+
+    @tailrec
+    def recursivelyParseState(state: ParseData): ParseData = state.parseState match {
+      case Default =>
+        val readString = readStringUntilNewline(inputStream)
+        if (readString == boundary) recursivelyParseState(state.changeState(AfterBoundary))
+        else state.done()
+      case AfterBoundary =>
+        val line = readStringUntilNewline(inputStream)
+        if (line.trim.isBlank) recursivelyParseState(state.changeState(AfterHeaderSpace))
+        else recursivelyParseState(state.addHeader(line))
+      case AfterParsedBody =>
+        val line = readStringUntilNewline(inputStream)
+        if (line.trim.isBlank) state.done()
+        else recursivelyParseState(state.addHeader(line).changeState(AfterBoundary))
+      case AfterHeaderSpace =>
+        val partInputStream = parseMultipartBodyPart(inputStream, boundaryBytes)
+        recursivelyParseState(state.completePart(partInputStream))
+    }
+
+    recursivelyParseState(ParseData(empty, List.empty, Default)).completedParts
+  }
+
+  private val TempFileThreshold = 52_428_800
```

Comment:
Should this threshold be configurable somehow?
I got this number from [http4s](https://github.com/http4s/http4s/blob/series/0.23/core/shared/src/main/scala/org/http4s/multipart/MultipartParser.scala#L656), but input is appreciated 😄

Reply:
So we'll read 52MB of data before falling back to a file? Seems quite a lot ;) We could make this configurable via `JdkHttpServerOptions`, but I'm not sure if it's worth the effort. A reasonable default might do for now.

Reply:
Doesn't seem to be too much work if i can pass it the same way `createFile` is passed through the `JdkHttpRequestBody` constructor?